src/hotspot/share/gc/g1/c2/g1BarrierSetC2.cpp

```diff
@@ -618,8 +620,8 @@ Node* G1BarrierSetC2::load_at_resolved(C2Access& access, const Type* val_type) c
   // SATB log buffer using the pre-barrier mechanism.
   // Also we need to add memory barrier to prevent commoning reads
   // from this field across safepoint since GC can change its value.
-  bool need_read_barrier = in_heap && (on_weak ||
-                                       (unknown && offset != top && obj != top));
+  bool need_read_barrier = (((on_weak || on_phantom) && !no_keepalive) ||
```

Comment:
I don't think we have any !in_heap && on_weak loads today. But if we did, they would indeed need read barriers.
We need read barrier if the the reference isn't provably strong... unless it's an AS_NO_KEEPALIVE access. That also reflects why the variable is called no_keepalive instead of keepalive; it is to reflect the shared decorator name used all over the place. I don't mind inverting it though, but personally found it easier to read when the names match our decorators.

Reply:
From this conversation the only change I can do is 'Turn (on_weak || on_phantom) into !on_strong'.
@fisk Is this correct? I am concern that it will include `unknown` decorator too.
I agree with Erik to keep !no_keepalive because he prefer it and this is code supported by GC group.

Reply:
Well if on_weak || on_phantom then it is provably a weak access. But I think the absence of the strong decorator does not prove it is weak, as it could have an unknown strength (via unsafe), in which case we need some extra logic to see if we can prove that an unknown strength access can't be weak.

Reply:
Okay. I will leave changes as it is.